Fix `bin`, `oct`, `dec` and `hex` when padding negative `BigInt`s

Tetralux opened the pull request:
As brought to my attention via @KlausC, there was a bug in my previous PR involving these functions, #18855, where negative numbers had the minus sign display after the padding and before the number.
~~This PR fixes that issue by printing an minus sign if appropriate, followed by printing out the absolute value of the number, and adds some more tests.~~
This PR adds a special method for padding `BigInt`s and then refactors `bin`, `oct`, `dec` and `hex` to simply call this new method. This replaces what I did #18855.
This new method simply calls the non-padding version of `base` with the same radix, handles the possible minus-sign, and pads the result.

Before this PR: `hex(-big(0x123), 5) == "0-123"`

After this PR: `hex(-big(0x123), 5) == "-00123"`

~~I also didn't remove some old tests that are more than covered by the tests from #18855. So I threw that into this PR as well - let me know if you'd rather have them in a different PR.~~

Tetralux committed:
Fix `bin`, `oct`, `dec`, `hex` for negative `BigInt` values

StefanKarpinski commented on base/gmp.jl:
```diff
@@ -521,8 +521,11 @@ hex(n::BigInt) = base(16, n)
     @eval function ($f)(n::BigInt, pad::Int)
         b = IOBuffer()
-        res = ($f)(n)
+        res = ($f)(abs(n))
```
This unfortunately creates another BigInt for `abs(n)`, which could be quite expensive. It would be preferable to use the GMP API to print `n` with padding without constructing `abs(n)`. One way to do it would be to temporarily construct `abs(n)` while sharing the same data pointer as `n`, but that may be a bit too tricky. Using GMP printing functions directly is probably the better approach.

KlausC replied:
> One way to do it would be to temporarily construct abs(n) while sharing the same data pointer as n, but that may be a bit too tricky.
couldn't we expect this tricks from the compiler? after all n is immuatable and abs(n) is local inside an 
expression. 

>Using GMP printing functions directly is probably the better approach.
which printing functions of GMP do you refer to? I only found get_...z_str, which does not support padding in the docu of current version 6.1.1.

StefanKarpinski replied:
The `BigInt` type is not actually immutable as that would prevent attaching a finalizer to it. BigInts are only immutable by convention – they do not support mutating operations. Ultimately, we'd like to change that precisely so that we can do optimizations like that, but it's not currently possible.

tkelman commented on test/bigint.jl:
```diff
@@ -305,11 +305,6 @@ end
-@test bin(big(3)) == "11"
-@test oct(big(9)) == "11"
-@test oct(-big(9)) == "-11"
-@test hex(big(12)) == "c"
```
Where else are the single-argument versions of these being tested? I'd rather leave these

H-225 replied:
Good call. Not sure why I thought these were cover by my other tests.

Tetralux committed:
Add special function for `base(radix, number, padding)`

  * Missed this from #18855; `base` needed a special method for
    `BigInt` with padding as well.

Tetralux committed:
Refactor `bin`, `oct`, `dec`, `hex` which pad the result

  ...to use the newly added special method for `base` for
  padding `BigInt`s. (See previous commit.)

H-225 commented:
Round two is ready for inspection. :)

StefanKarpinski commented on base/gmp.jl:
```diff
@@ -518,25 +518,33 @@ oct(n::BigInt) = base( 8, n)
+        s = s[2:end]
+        write(buf, '-')
+    end
+    if length(s) < pad
```
`sizeof(s)` would be better here since `length(s)` is O(n) in string length and we know this has to be ASCII. Also, isn't the `if` redundant? The `for` loop will execute zero time if `length(s) < pad` anyway.

StefanKarpinski reviewed (changes requested):
This approach is good. It could probably be made even faster, but that kind of additional optimization can wait until a future time when someone really needs it.

Tetralux committed:
Some easy optimizations

  * Remove redundant if.
  * Use `sizeof` to determine string length since it's guarenteed
    to be ASCII (thereby making it O(1)).

H-225 commented:
...and round three.

H-225 commented:
Bump?

StefanKarpinski commented:
Thanks for fixing this in the first place and for seeing it through.

H-225 commented:
Pleasure.

H-225 commented:
Also, for future reference, this PR also fixed `base(b, n, pad)`, which called `sizeof` on `n` which I missed from #18855.

tkelman commented:
these methods were added by #18855 which isn't on release-0.4 or release-0.5, so not a backport candidate unless done along with that
